### ISSUE TITLE
How to scale TFRS?

### ISSUE THREAD

**Issue by dgoldenberg-audiomack** [tensorflow/recommenders#201]
The documentation does not appear to offer much guidance on how to scale up a TFRS based solution, for example, here: https://www.tensorflow.org/recommenders/examples/basic_retrieval.

To start, let's consider a (common, I'm sure) scenario of N million users and M million items (such as movies). We would be envisioning at least millions of feature records also, but for the moment we could set that aside and contemplate just the 'basic retrieval' use-case/example.

There are at least 3 considerations while designing an actual recommender, in terms of scale:

1. How to feed input data into TFRS. I'm starting with Parquet from our datalake. Due to the size of this data and various filtering and augmentation of data sets, I'm looking at having to read Parquet in a Spark job, transforming the data, then persisting it into an intermediary Parquet format ready to be consumed by TFRS. Issue: extra I/O. I don't see a way to do all the transformations in TF/TFRS, especially joins.
2. The building of vocabularies. From the example:
```
unique_movie_titles = np.unique(np.concatenate(list(movie_titles)))
unique_user_ids = np.unique(np.concatenate(list(user_ids)))
```
I can uniquefy my data in Spark when prepping Parquet for TFRS. However, it appears that building a model in TFRS is oriented toward using numpy arrays:
```
user_model = tf.keras.Sequential([
  tf.keras.layers.experimental.preprocessing.StringLookup(
      vocabulary=unique_user_ids, mask_token=None),
  # We add an additional embedding to account for unknown tokens.
  tf.keras.layers.Embedding(len(unique_user_ids) + 1, embedding_dimension)
])
```
How are vocabularies expected to perform with millions of ID's? how will this affect memory and if this is not expected to scale, what alternatives are there to vocabulary/numpy array approach? 
3. Getting the recommendations out at scale. In my use-case, I'm going to ask the model to generate more recommendations than actually needed, so that I can perform some post-filtering based on a set of a few rules. I need a scalable way to convert tensors en masse to Spark datasets which will write Parquet or CSV back into the datalake. This may mean extracting recs as Parquet into intermediary files then having separate code to turn that into the final, post-processed results. Issue: extra I/O and processing.

All in all, what this feels like is potentially a need for a tight integration with Spark datasets where it is easy and seamless to load TF datasets from Spark and then just as easy to convert the resulting recommendations (e.g. a `tfrs.layers.factorized_top_k.BruteForce` index) to Spark datasets.

So far, I don't seem to see a way to do this. The LinkedIn's [Spark-TFRecord library](https://github.com/linkedin/spark-tfrecord) and the [Spark Tensorflow Connector](https://github.com/tensorflow/ecosystem/tree/master/spark/spark-tensorflow-connector) both require intermediary tfrecord files.  I'm "fishing" for a more Spark oriented way to build a TFRS model and then extract results en masse for millions of input users.

**Comment by maciejkula** [comment]
We don't offer specific guidance because everything that applies to scaling TensorFlow models applies here equally.

1. Preprocessing data. Here, a Spark job looks sensible. You could also have a look at [TensorFlow Transform](https://www.tensorflow.org/tfx/transform/get_started) which uses Apache Beam for parallelizing transformations.
2. Building of vocabularies. The [`StringLookup`](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/preprocessing/StringLookup) can accept either a numpy array or a path to a vocabulary file; either way millions of ids will be fine.
3. Getting the recommendations out. For fast prediction I'd suggest using the [ScaNN](https://www.tensorflow.org/recommenders/examples/efficient_serving) approach. It should be able to very efficiently generate top-K predictions you will then be able to post-process using Spark.

I'm afraid I know very little about Spark, but the basic approach of generating training data using Spark, and post-processing results using Spark again makes sense to me. I imagine you'd also be able to load the trained model into memory in your Spark functions and call it, so (3) could in principle be done entirely in Spark.

**Comment by dgoldenberg-audiomack** [comment]
Thanks, @maciejkula. This makes sense to me; I'll update the ticket as I experiment with larger datasets.

This is possibly a "naive" intuitive sense, however, "wouldn't it be nice" if TensorFlow's datasets could "ride" Spark datasets directly. It's probably wishful thinking but if there was a way to integrate Spark datasets directly into TF, as in, some TF dataset class being a subclass of Spark dataset, the framework would avoid having to deal with potentially massive amounts of intermediary files (on the input and on the output).

In other words, a TensorFlow dataset could become enormously powerful if it WERE a Spark dataset, with all the rich functionality it offers, including joins etc.etc.

**Comment by MLnick** [comment]
@dgoldenberg-audiomack I agree, the experience is far from seamless currently.

The closest looking thing is actually: https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/orca.html (from the BigDL team at Intel I think).

I plan to give it a go on YARN to see how it works!

**Comment by MLnick** [comment]
I’d like to try scaling out a basic retrieval model using TF recommenders,
to see how it can scale up to larger data size.

I think user/item ids would be overall somewhat limited in size because it
would still need to fit into single node memory, but still should be
possible to have a few million.

**Comment by dgoldenberg-audiomack** [comment]
Nice. I think it would be great to have a full example of how to scale TFRS so it's easy to take and plop into AWS EMR or the like with few tweaks.
